### PR TITLE
feat: add validation_frequency option to fine-tuning

### LINKED ISSUE #811: Add validation options to fine tuning
### Describe the workflow you want to enable

A few things that would be nice to have for finetuning:

- option to specify disabling validation
- option to specify frequency of validation
- TBC

### Describe your proposed solution

Implement options

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### Impact

None

### PR DESCRIPTION
**Pull request by Hemanth-1354**
## Issue
Closes #811

## Motivation and Context
Fine-tuning currently validates every epoch unconditionally. For long runs
or large validation sets this adds overhead, and there's no way to disable
validation (and the early stopping that depends on it) when you don't need it.

This adds a `validation_frequency: int | None = 1` parameter to
`FinetunedTabPFNBase` (and threads it through `FinetunedTabPFNClassifier` /
`FinetunedTabPFNRegressor`):

- `validation_frequency=1` (default): validates every epoch, identical to
  current behavior.
- `validation_frequency=N`: validates (and checks early stopping) only every
  N epochs.
- `validation_frequency=None`: skips validation for the entire run, which
  also disables early stopping. If `early_stopping=True` is passed alongside
  `validation_frequency=None`, a `UserWarning` is raised since the flag would
  otherwise silently have no effect.

Skipped epochs report `NaN` as the validation metric, reusing the existing
`not np.isnan(...)` guards already in the checkpoint-saving and
early-stopping code paths. Early stopping's restore-to-best-weights logic is
gated on a derived `effective_early_stopping` flag rather than the raw
`self.early_stopping`, so a run with `validation_frequency=None` can't
accidentally revert to the pre-training weights at the end.

## Public API Changes
- [ ] No Public API changes
- [x] Yes, Public API changes (Details below)

New optional keyword-only parameter `validation_frequency` on
`FinetunedTabPFNBase.__init__`, `FinetunedTabPFNClassifier.__init__`, and
`FinetunedTabPFNRegressor.__init__`. Defaults to `1`, which preserves
existing behavior exactly, no breaking changes.

## How Has This Been Tested?
- Added `test__finetuned_tabpfn_classifier__validation_frequency_skips_epochs`:
  mocks `_evaluate_model` and asserts it's called only on the expected
  epochs (plus the one pre-loop baseline eval) when `validation_frequency=2`.
- Added `test__finetuned_tabpfn_classifier__validation_frequency_none_disables_es`:
  asserts the `UserWarning` fires for `early_stopping=True` +
  `validation_frequency=None`, and that fine-tuned weights are *not* reverted
  to the base model despite `early_stopping=True`, proving early stopping is
  actually disabled rather than just skipped-but-still-restoring.
- `ruff check` / `ruff format --check` pass on all changed files.
- Existing finetuning test suite is unaffected (default `validation_frequency=1`
  reproduces the prior code path exactly).

## Checklist
- [x] The changes have been tested locally.
- [x] Documentation has been updated (if the public API or usage changes).
- [x] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
- [x] The code follows the project's style guidelines.
- [x] I have considered the impact of these changes on the public API.

### CODE REVIEW

**Comment by anuragg1209** [comment]
Hi @Hemanth-1354, thanks for the PR — well-thought-out description and tests, and `validation_frequency` is a feature we'd like to have!

The PR needs a rebase: **#1101** landed after you opened this and reworked `finetuned_base.py` (validation gating, checkpointing, early stopping), which is where the conflicts come from. It also means `main` now already supports **disabling validation entirely** via `validation_split_ratio=None` (or `0`) — trains on all data, skips per-epoch eval, and auto-disables early stopping — so the `validation_frequency=None` case here is now redundant.

After rebasing, we'd suggest narrowing the scope to the missing part — **validate every N epochs** — which also shrinks the diff:

- `validation_frequency: int = 1` (positive int only, no `None`) — drops the `None` type-checks and the `UserWarning` interaction with `early_stopping`.
- `main` now has the seams you need (`do_validation`, `early_stopping_enabled`, an else-branch logging train loss), so the integration reduces to one derived flag at the per-epoch eval site:

  ```python
  run_validation = do_validation and (epoch + 1) % self.validation_frequency == 0
  ```

- Your NaN-sentinel for skipped epochs still fits: the early-stopping block is already guarded by `not np.isnan(primary_metric)`, so patience won't tick on skipped epochs. Please keep your change moving the NaN check into `is_best` — interval checkpoints are for crash recovery/resume and don't need a validation metric, so they should still save on skipped epochs.
- On skipped epochs, the checkpoint `metrics` dict shouldn't contain the NaN eval result — `main` already has a fallback to `{"train_loss": mean_train_loss}` for the no-validation case, so just gate that branch on `run_validation` instead of `do_validation`.
- Small docs note: with `validation_frequency > 1`, `early_stopping_patience` counts validation runs, not epochs.

Your `validation_frequency=2` test should carry over as-is; the `None` test can be dropped along with that case.

Sorry for the moving target — the timing with #1101 was unlucky. Happy to help if anything is unclear during the rebase, and thanks again for contributing!

**Comment by Hemanth-1354** [comment]
Thanks for the detailed review @anuragg1209! I've rebased on main and narrowed the scope as suggested:

- Dropped `validation_frequency=None` - now just `validation_frequency: int = 1`
- Simplified integration using main's `do_validation` / `early_stopping_enabled` seams
- One-line flag: `run_validation = do_validation and (epoch + 1) % self.validation_frequency == 0`
- NaN sentinel moved into `is_best` so interval checkpoints save on skipped epochs
- Checkpoint metrics gated on `run_validation` instead of `do_validation`
- Added positive integer validation
- Dropped the `None` test, kept `validation_frequency=2` test

@anuragg1209

**Comment by anuragg1209** [comment]
Hi @Hemanth-1354, thanks again for proposing validation_frequency and for adapting the implementation as the fine-tuning code evolved. We retained the core scheduling approach and moved it forward on top of the current main. The main differences are that validation disabling continues to use the existing validation_split_ratio=None/0 API, validation frequency is validated at fit time as a positive integer, and we added resume/checkpoint/early-stopping safeguards and broader test coverage. 

We’ll acknowledge your contribution in the PR, so we added you as a co-author on the feature commit to ensure your contribution is recorded in the Git history.

We'll close this PR in favor of #1144, of which you are also a co-author. Thanks!
